API: more reliable, descriptive VimL errors

justinmk opened the pull request:
- [x] Capture *abort-causing non-exception* errors.
  - `nvim_command` did this correctly already, because `do_cmdline()` sets `msg_list`. That same pattern is needed in `nvim_eval` and `nvim_call_function`, else such errors are not reported as errors to API clients.
- Return VimL errors instead of generic errors for:
  - [x] nvim_call_function
  - [x] nvim_call_dict_function
  - [x] nvim_eval
- [x] Fix tests which were silently broken before this change.

This violates #6150 (because VimL errors _are_ localized) where we agreed not to localize API errors.  But that can be fixed later.

This should make `nvim_call_{dict_}function()` and `nvim_eval()` behave more correctly, as well as giving better error messages for VimL errors. "More correctly" because previously they did not fail for some types of errors.

@garyburd This should fix the weirdness [you reported](https://gitter.im/neovim/neovim?at=5a39a3cc00d569114bc7e95a).

justinmk committed:
doc

justinmk committed:
msgpack.vim: fix syntax errors, python2 errors

justinmk committed:
API: return non-generic VimL errors

- Return VimL errors instead of generic errors for:
  - nvim_call_function
  - nvim_call_dict_function
- Fix tests which were silently broken before this change.

This violates #6150 where we agreed not to translate API errors.  But
that can be fixed later.

justinmk commented on src/nvim/message.c:
```diff
@@ -504,6 +505,9 @@ int emsg(const char_u *s_)
         did_emsg = true;
+        if (msg_first_ignored_err == NULL) {
+          msg_first_ignored_err = xstrdup(s);
+        }
```
This is ugly, but there's no way around it that I could find. `call_func()` does not treat _non-exception, abort-triggering_ `emsg()` errors the same way as `do_cmdline_cmd()`, and following all of the global flags is a nightmare.

So this is a special-case global that holds the error in this very specific case.

justinmk commented on runtime/autoload/msgpack.vim:
```diff
@@ -62,10 +63,10 @@ function s:msgpack_init_python() abort
               \. "    return ('{\"_TYPE\": v:msgpack_types.integer,'\n"
               \. "            + '\"_VAL\": [{sign},{v1},{v2},{v3}]}').format(\n"
-              \. "              sign=1 if timestamp >= 0 else -1,\n"
+              \. "              sign=1 if timestamp >= 0 else sign=-1,\n"
```
@ZyX-I I took a guess at how to fix this, can you help? This is the failing test:

```
[  ERROR   ] /home/vagrant/neovim/test/functional/plugin/shada_spec.lua @ 2077: autoload/shada.vim function shada#get_binstrings works
stack traceback:
        test/functional/helpers.lua:99: in function 'shada#get_binstrings'
        /home/vagrant/neovim/test/functional/plugin/shada_spec.lua:2069: in function 'getbstrings_eq'
        /home/vagrant/neovim/test/functional/plugin/shada_spec.lua:2083: in function </home/vagrant/neovim/test/functional/plugin/shada_spec.lua:2077>
```

ZyX-I replied:
This is kwarg("sign", ternary(cond=ge(var("timestamp"), num("0")), if_true=num("1"), if_false=num("-1"))), it is Python ternary operator used to determine value of a keyword argument. What you have here now is not going to work, not only assignment is not an expression, *there is no assignment here*. And I do not think this is full error message.

justinmk replied:
@ZyX-I Yes this doesn't work, I did not want to continue debugging it until getting your opinion. Do we even need `s:msgpack_init_python()` anymore?

justinmk replied:
@ZyX-I what did you intend with `timestamp = int(timestamp.timestamp())` ? `timestamp()` function does not exist on that object in python2 nor python3.

justinmk commented on test/functional/lua/luaeval_spec.lua:
```diff
@@ -51,12 +51,12 @@ describe('luaeval()', function()
     it('are successfully transformed', function()
-      funcs.luaeval('rawset(_G, "d", {})')
-      funcs.luaeval('rawset(d, "d", d)')
+      command('lua rawset(_G, "d", {})')
```
@ZyX-I The `funcs.luaeval` variant gives this error:

```
[ RUN      ] luaeval() recursive lua values are successfully transformed: ERR
test/functional/helpers.lua:99: E5101: Cannot convert given lua type
```

ZyX-I replied:
Apparently it used to give the same error before, just ignored that.

justinmk committed:
API: better way to capture abort-causing non-exception errors

This condition is not perfectly reliable:
    (did_emsg && force_abort && !current_exception)

The more proper way to check for abort-causing non-exception errors is
to set up `msg_list` using the "pattern" given by do_cmdline().

justinmk committed:
API: nvim_eval(): return non-generic VimL errors

Use the same pattern as nvim_call_function (_call_function).

justinmk committed:
test: API: fix tests after improved error capture

justinmk committed:
nvim_eval: fix memory leak

justinmk commented:
@ZyX-I changes to msgpack.vim seem to work on unix, but appveyor still fails with the similar errors I had been seeing:

```
  [  ERROR   ] C:/projects/neovim/test/functional\plugin\msgpack_spec.lua @ 284: autoload/msgpack.vim function msgpack#strptime works
  test\functional\helpers.lua:99: Vim(return):error caught in request handler 'python_eval ('shada_dict_strptime()',)':
  
  stack traceback:
  	test\functional\helpers.lua:99: in function 'nvim_eval'
  	C:/projects/neovim/test/functional\plugin\msgpack_spec.lua:287: in function <C:/projects/neovim/test/functional\plugin\msgpack_spec.lua:284>
  
  [  ERROR   ] C:/projects/neovim/test/functional\plugin\shada_spec.lua @ 1361: autoload/shada.vim function shada#strings_to_sd works with multiple items
  test\functional\helpers.lua:99: Vim(return):error caught in request handler 'python_eval ('shada_dict_strptime()',)':
  
  stack traceback:
  	test\functional\helpers.lua:99: in function 'request'
  	test\functional\helpers.lua:213: in function 'nvim_command'
  	C:/projects/neovim/test/functional\plugin\shada_spec.lua:1340: in function 'strings2sd_eq'
  	C:/projects/neovim/test/functional\plugin\shada_spec.lua:1362: in function <C:/projects/neovim/test/functional\plugin\shada_spec.lua:1361>
```

Going to skip that test on Windows for now, seems like this feature never worked.

justinmk committed:
msgpack.vim: require python3 on Windows

timestamp.strftime('%s') workaround only works on unix.

ref: https://github.com/neovim/neovim/pull/8371#discussion_r186311766
